Support for cloud mode in Google Analytics 4 (#532)
rudderlabs/rudderstack-docs: destinations/analytics/google-analytics-4/google-analytics-4-cloud-mode.mdx

    <th>Google Analytics 4</th>
    <th>RudderStack</th>
    <th>Google Analytics 4</th>
  </tr>
  <tr>
    <td>Products Searched</td>
    <td><code class=“inline-code”>search</code></td>
    <td><code class=“inline-code”>query</code></td>
    <td><code class=“inline-code”>search_term</code></td>
  </tr>
  <tr>
    <td>Product List Viewed</td>
    <td><code class=“inline-code”>view_item_list</code></td>
    <td rowspan="2"><code class=“inline-code”>list_id</code></br><code class=“inline-code”>category</code></td>
    <td rowspan="2"><code class=“inline-code”>item_list_id</code></br><code class=“inline-code”>item_list_name</code></td>
  </tr>
  <tr>
   <td>Product Clicked</td>
    <td><code class=“inline-code”>select_item</code></td>
  </tr>
  <tr>
    <td>Promotion Viewed</br>Promotion Clicked</td>
    <td><code class=“inline-code”>view_promotion</code></br><code class=“inline-code”>select_promotion</code></td>
    <td><code class=“inline-code”>position</code></td>
    <td><code class=“inline-code”>location_id</code></td>
  </tr>
   <tr>
    <td>Product Viewed</br>Product Added</br>Product Removed</br>Cart Viewed</br>Checkout Started</br>Product Added to Wishlist</td>

>     <td><code class=“inline-code”>view_item</code></br><code class=“inline-code”>add_cart</code></br><code class=“inline-code”>remove_from_cart</code></br><code class=“inline-code”>view_cart</code></br><code class=“inline-code”>begin_checkout</code></br><code class=“inline-code”>add_to_wishlist</code></td>

Reviewer: ```suggestion
    <td><code class=“inline-code”>view_item</code></br><code class=“inline-code”>add_to_cart</code></br><code class=“inline-code”>remove_from_cart</code></br><code class=“inline-code”>view_cart</code></br><code class=“inline-code”>begin_checkout</code></br><code class=“inline-code”>add_to_wishlist</code></td>
```

| RudderStack    | Google Analytics 4  |
| :-------------| :-----------------|
| product_id         | `item_id`        |
| name               | `item_name`      |
| coupon               | `coupon`      |
| price              | `price`          |
| position           | `index`          |
| category           | `item_category`  |
| brand           | `item_brand`  |
| variant           | `item_variant`  |
| quantity           | `quantity`  |

### Rules for naming custom events

You must follow the below rules while choosing a name for the custom events:

- Event names are case sensitive. For example, `my_event` and `My_Event` are two distinct events.
- Event names must start with a letter. Only letters, numbers, and underscores are permitted. **DO NOT** use spaces.
- Do not use reserved prefixes and event names. Refer to the <a href="#faq">FAQ</a> section to know more about the reserved event names and prefixes.

## FAQ

### What might be the reason for an error message from Google Analytics?

> Google Analytics 4 has some reserved prefixes and events names which cannot be used. The list of such prefixes is mentioned below:

Reviewer: Lets seperate this

- _ (underscore)
- firebase_
- ga_
- google_

> - gtag.

Reviewer: These are for custom events

Reviewer: We can provide [this](https://support.google.com/analytics/answer/10085872#zippy=%2Creserved-prefixes-and-event-names%2Cweb) link for custom events.

> For the list of reserved event names, refer to the [Reserved Names](https://developers.google.com/analytics/devguides/collection/protocol/ga4/reference?client_type=gtag#reserved_names) guide.

Reviewer: This link provided here are for GA4 standard events

A sample `track` call is as shown:

```javascript
rudderanalytics.track('product list viewed', {
  list_id: "related_products",
  category: "Related_products",
  products: [{
    product_id: "507f1f77bcf86cd799439011",
    name: "Monopoly: 3rd Edition",
    coupon: "SUMMER_FUN",
    category: "Apparel",
    brand: "Google",
    variant: "green",
    price: "19",
    quantity: "2",
    position: "1",
    affiliation: "Google Merchandise Store",
    currency: "USD",
    discount: 2.22,
    item_category2: "Adult",
    item_category3: "Shirts",
    item_category4: "Crew",
    item_category5: "Short sleeve",
    item_list_id: "related_products",
    item_list_name: "Related Products",
    location_id: "L_12345"
  }]
});
```

## E-commerce

RudderStack supports e-commerce tracking for Google Analytics 4. You can use the [RudderStack e-commerce specification](https://rudderstack.com/docs/rudderstack-api/api-specification/rudderstack-ecommerce-events-specification/) for sending events while instrumenting your site with the RudderStack SDK.

> The below table lists the mappings of the `track` events and properties that are passed to Google Analytics 4 e-commerce events and properties:

Reviewer: ```suggestion
The below table lists the mappings of the `track` events and properties that are passed to Google Analytics 4 events and properties:
```
GA4 docs have not mentioned about ecommerce event [Ref](https://developers.google.com/analytics/devguides/collection/protocol/ga4/reference/events)

    <td><code class=“inline-code”>cart</code></td>
  </tr>
   <tr>
    <td>Group</td>
    <td><code class=“inline-code”>join_group</code></td>
    <td>-</td>
    <td>-</td>
  </tr>
  <tr>
    <td>Payment Info Entered</td>
    <td><code class=“inline-code”>add_payment_info</code></td>
    <td><code class=“inline-code”>payment_method</code></td>
    <td><code class=“inline-code”>payment_type</code></td>
  </tr>
  <tr>
    <td>Payment Info Entered</td>
    <td><code class=“inline-code”>add_shipping_info</code></td>
    <td><code class=“inline-code”>shipping_method</code></td>
    <td><code class=“inline-code”>shipping_tier</code></td>
  </tr>
</table>

<div class="infoBlock">

The <strong>Payment Info Entered</strong> event is mapped on the basis of parameters passed to it:
<ul>
  <li>If <code class=“inline-code”>payment_method</code> is passed along with the other fields, it maps to <code class=“inline-code”>add_payment_info</code></li> event.
 <li>If <code class=“inline-code”>shipping_method</code> is passed along with the other fields, it maps to <code class=“inline-code”>add_shipping_info</code></li> event.
  </ul>
</div>

<div class="infoBlock">

> There must be an <code class="inline-code">id</code> and <code class="inline-code">name</code> associated for each product in the order. Refer to the <a href="https://developers.google.com/gtagjs/reference/ga4-events#view_item_list">Item list</a> for more information on Google Analytics 4 e-commerce events and corresponding properties.

Reviewer: ```suggestion
There must be an <code class="inline-code">id</code> and <code class="inline-code">name</code> associated for each product in the order. Refer to the <a href="https://developers.google.com/gtagjs/reference/ga4-events#view_item_list">Item list</a> for more information on Google Analytics 4 events and corresponding properties.
```
The item list referenced in the url is specific to view_item_list event.

> There must be an <code class="inline-code">id</code> and <code class="inline-code">name</code> associated for each product in the order. Refer to the <a href="https://developers.google.com/gtagjs/reference/ga4-events#view_item_list">Item list</a> for more information on Google Analytics 4 e-commerce events and corresponding properties.

Reviewer: The statement is quite unclear. Do we mean `item_id` and `item_name` must be present inside Items parameters

Reviewer: @utsabc any suggestion here. This statement is part of our device mode doc.

Reviewer 2: Do we support items array or product array. If we are mapping from product array we should mention that

Reviewer 2: We should be clear about what structure we are expecting from user and to what structure we are mapping

</div>

Most of the above mentioned events include [`items`](https://developers.google.com/analytics/devguides/collection/protocol/ga4/reference/events#view_item_list) parameter which accepts an `Item` array. The below table details out the mappings for `Items` array:

| RudderStack    | Google Analytics 4  |
| :-------------| :-----------------|
| product_id         | `item_id`        |
| name               | `item_name`      |
| price              | `price`          |
| position           | `index`          |
| category           | `item_category`  |
| list_id            | `item_list_id`   |

> | category           | `item_list_name` |

Reviewer: There are a few more mapping. Let's update this

### Rules for naming events

> You must follow the below rules while choosing a name for the new events:

Reviewer: This applies to custom events only. [Ref](https://support.google.com/analytics/answer/10085872#zippy=%2Creserved-prefixes-and-event-names%2Cweb%2Cin-this-article)

    <td><code class=“inline-code”>unlock_achievement</code></td>
    <td><code class=“inline-code”>properties.achievement_id</code></td>
    <td><code class=“inline-code”>achievement_id</code></td>
  </tr>
    <tr>
    <td>view_search_results</td>
    <td><code class=“inline-code”>view_search_results</code></td>
    <td><code class=“inline-code”>properties.search_term</code></td>
    <td><code class=“inline-code”>search_term</code></td>
  </tr>
</table>

### Rules for naming custom events

You must follow the below rules while choosing a name for the custom events:

- Event names are case sensitive. For example, `my_event` and `My_Event` are two distinct events.
- Event names must start with a letter. Only letters, numbers, and underscores are permitted. **DO NOT** use spaces.
- Do not use reserved prefixes and event names. Refer to the <a href="#faq">FAQ</a> section to know more about the reserved event names and prefixes.

## FAQ

### What are the reserved prefixes in Google Analytics 4?

The reserved prefixes in Google Analytics 4 cannot be used for custom events. The list of such prefixes is mentioned below:

- _ (underscore)
- firebase_
- ga_
- google_
- gtag.

> Refer to the [Google document](https://support.google.com/analytics/answer/10085872#zippy=%2Cweb%2Cin-this-article%2Creserved-prefixes-and-event-names) for more information.

Reviewer: ```suggestion
Refer to the [GA4 document](https://support.google.com/analytics/answer/10085872#zippy=%2Cweb%2Cin-this-article%2Creserved-prefixes-and-event-names) for more information.
```

> ### What are the reserved event names in Google Analytics 4?

Reviewer: ```suggestion
### What are the reserved event, parameter, user property names in Google Analytics 4?
```

> Google Analytics 4 has some reserved events names which cannot be used. Refer to the [Reserved names](https://developers.google.com/analytics/devguides/collection/protocol/ga4/reference?client_type=gtag#reserved_names) for complete list of reserved event names.

Reviewer: ```suggestion
Google Analytics 4 has some reserved event, parameter, and user property names that cannot be used. Refer to the [Measurement Protocol (Google Analytics 4)](https://developers.google.com/analytics/devguides/collection/protocol/ga4)](https://developers.google.com/analytics/devguides/collection/protocol/ga4/reference?client_type=gtag#reserved_names) for complete list of reserved event names.
```

> Google Analytics 4 has some reserved event, parameter, and user property names that cannot be used. Refer to the [Measurement Protocol (Google Analytics 4)](https://developers.google.com/analytics/devguides/collection/protocol/ga4)](https://developers.google.com/analytics/devguides/collection/protocol/ga4/reference?client_type=gtag#reserved_names) guide for a complete list of reserved names.

Reviewer: ```suggestion
Google Analytics 4 has some reserved event, parameter, and user property names that cannot be used. Refer to the [Measurement Protocol (Google Analytics 4)](https://developers.google.com/analytics/devguides/collection/protocol/ga4/reference?client_type=gtag#reserved_names) guide for a complete list of reserved names.
```

</div>

Most of the above mentioned events include [`items`](https://developers.google.com/analytics/devguides/collection/protocol/ga4/reference/events#view_item_list) parameter which accepts an `Item` array. The below table details out the common mappings for `Items` array:

| RudderStack    | Google Analytics 4  |
| :-------------| :-----------------|
| product_id         | `item_id`        |
| name               | `item_name`      |
| coupon               | `coupon`      |
| price              | `price`          |
| position           | `index`          |
| category           | `item_category`  |
| brand           | `item_brand`  |
| variant           | `item_variant`  |
| quantity           | `quantity`  |

The below mentioned e-commerce events include [`items`](https://developers.google.com/analytics/devguides/collection/protocol/ga4/reference/events#view_item_list) parameter which accepts an `Item` array. 

| RudderStack    | Presence of `items` parameter |
| :-------------| :-----------------|
| Product List Viewed         | Required        |

> | Product Clicked         | Optional      |

Reviewer: ```suggestion
| Product Clicked         | Required      |
```

| Product Viewed              | Required      |
| Product Added              | Required         |
| Product removed           | Required          |
| Cart Viewed           | Required  |
| Checkout Started           | Required  |
| Payment Info entered           | Required  |
| Order Completed           | Required  |
| Order Refunded           | Optional  |
| Product Added to Wishlist           | Required  |

> | View Search Result           | Optional  |

Reviewer: ```suggestion
| View Search Results           | Optional  |
```
